crates/sdk/src/tx.rs

```diff
@@ -3026,13 +3077,43 @@ pub async fn build_shielded_transfer<N: Namada>(
         args.tx_code_path.clone(),
         data,
         add_shielded_parts,
-        fee_amount,
+        fee_per_gas_unit,
         &signing_data.fee_payer,
     )
     .await?;
     Ok((tx, signing_data))
 }
 
+// Check if the transaction will need to pay fees via the masp and extract the
+// right masp data
+async fn get_masp_fee_payment_amount<N: Namada>(
+    context: &N,
+    args: &args::Tx<SdkTypes>,
+    fee_amount: DenominatedAmount,
+    fee_payer: &common::PublicKey,
+    gas_spending_keys: Vec<ExtendedSpendingKey>,
+) -> Result<Option<MaspFeeData>> {
+    let fee_payer_address = Address::from(fee_payer);
+    let balance_key = balance_key(&args.fee_token, &fee_payer_address);
+    let balance = rpc::query_storage_value::<_, token::Amount>(
+        context.client(),
+        &balance_key,
+    )
+    .await
+    .unwrap_or_default();
+    let total_fee = checked!(fee_amount.amount() * u64::from(args.gas_limit))?;
+
+    Ok(match total_fee.checked_sub(balance) {
```

Comment:
nit: I think it's clearer to check `total_fee > balance` and then calculate the diff afterwards